### PR TITLE
render should always be called regardless of shouldComponentUpdate & fix render arity

### PR DESCRIPTION
**Pull request by ghost**
To match with [react lifecycle](https://facebook.github.io/react/docs/react-component.html)

also includes fixes to other tests

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,6 +25,9 @@
     "controller"
   ],
   "author": "Boris Cherny <boris@performancejs.com>",
+  "contributors": [
+    "Chris Khoo"
```
Comment (bcherny) [inline package.json]: Add email so people know who to bug? :)

**Review by bcherny — APPROVED** [review]
LGTM. If we're aiming for React-compatability, we might also want to update `render`'s arity.

**Comment by ghost** [comment]
Ahh yes, if the file doesn't exist (as it doesn't in CI), it'll fail.  Hence forcing with `-f` will always exit with 0 code but either way works.

File: package.json [hunk 2]
```diff
@@ -6,9 +6,9 @@
   "main:esnext": "index.es2015.js",
   "typings": "index.d.ts",
   "scripts": {
-    "build": "rm index.d.ts && tsc -d -m es2015 --outDir . && mv ./index.js ./index.es2015.js && tsc -m commonjs -t es5  --outDir .",
+    "build": "rm -f index.d.ts && tsc -d -m es2015 --outDir . && mv ./index.js ./index.es2015.js && tsc -m commonjs -t es5  --outDir .",
```
Comment (bcherny) [inline package.json]: Something about `-f` always makes me nervous :)

File: test.ts [hunk 3]
```diff
@@ -228,8 +261,32 @@ describe('Component', () => {
...
+        bindings: {
+          a: '<',
+          b: '<'
+        },
+        controller,
+        template: `{{a}}`
+      })
+
+    const el = element('<my-component a="a" b="b"></my-component>')
+    const parentScope = extend($rootScope.$new(true), {
```
Comment (ghost) [inline test.ts]: _.extend is required because the chrome version used in circle is apparently quite old

File: index.ts [hunk 4]
```diff
@@ -17,50 +17,47 @@ abstract class NgComponent<Props, State> {
   */
   // nb: this method is explicity exposed for unit testing
   public $onChanges(changes: {}) {
+    console.info('TODOCK onChanges')
```
Comment (bcherny) [inline index.ts]: rm

File: index.ts [hunk 5]
```diff
@@ -17,50 +17,47 @@ abstract class NgComponent<Props, State> {
...
     this.componentWillMount()
   }
 
   $postLink() {
+    console.info('TODOCK componentDidMount')
     this.componentDidMount()
   }
 
   $onDestroy() {
+    console.info('TODOCK componentWillUnmount')
```
Comment (bcherny) [inline index.ts]: rm consoles

File: test.ts [hunk 6]
```diff
@@ -1,51 +1,60 @@
...
+      spyOn(a, 'render')
 
       // call #1
       a.$onChanges({
         a: {currentValue: 42, previousValue: undefined, isFirstChange: () => true},
         b: {currentValue: undefined, previousValue: undefined, isFirstChange: () => true}
       })
-      expect(spy.calls.mostRecent().args[0])
-        .toEqual({ a: 42, b: undefined })
+      expect(a.props).toEqual({ a: 42, b: undefined })
```
Comment (bcherny) [inline test.ts:25]: Nice - is TS ok with this?

File: test.ts (hunk 3 above)
Comment (bcherny) [reply test.ts]: _.assign?

File: index.ts (hunk 5 above)
Comment (ghost) [reply index.ts]: trying to fix a problem

image

componentWillMount & componentDidMount gets called after onChange event (e. where render gets called)... seriously out of order

File: test.ts (hunk 6 above)
Comment (ghost) [reply test.ts:25]: yup

File: index.ts (hunk 5 above)
Comment (ghost) [reply index.ts]: seems to me if $onChanges always gets called at the start, we could just put componentDidMount/WillMount in $onChanges?

what do u think?

Comment (bcherny) [reply index.ts]: Would that change the semantics in any cases from the consumer's point of view?

**Pull request merged**